docs/enhancements/fep-001-api-fleet-templating.md

```diff
@@ -0,0 +1,295 @@
+# EP: Definition of the Fleet and Device API spec
+<!-- this format is inspired by the K8S KEP format https://raw.githubusercontent.com/kubernetes/enhancements/master/keps/NNNN-kep-template/README.md -->
+## Summary
+
+The Fleet API is designed to manage big groups of devices that share configuration
+and operating system image.
+
+Some use cases point to the necessity of customizing some of those details based
+on the device's metadata.
+
+## Motivation
+
+Allowing templating of the device spec in the fleet API we could achieve a more
+flexible management, allowing customers to manage devices in a more granular way.
+
+### Goals
+
+* Define templating mechanism for the device spec template configuration in fleets,
+  based on labels and other metadata.
+
+* Enable different os images based on the device labels.
+```yaml
+  template:
+    spec:
+      os:
+        image: access-control-{{ .device.labels["region"] }}:1.0.0
+```
+
+* Enable different references for a config source based on labels:
+```yaml
+- name: microshift-manifests
+  configType: GitConfigProviderSpec
+  gitRef:
+    repository: basic-nginx-demo
+    targetRevision: 1.0.0-{{ .device.labels["region"] }}
+    path: /basic-nginx-demo/configuration
+```
+
+* Enable different folders within a config source based on labels:
+
+```yaml
+- name: microshift-manifests
+  configType: GitConfigProviderSpec
+  gitRef:
+    repository: basic-nginx-demo
+    targetRevision: 1.0.0
+    path: /basic-nginx-demo/configuration/{{ .device.labels["region"] }}
+```
+
+* Enable different folders within a config source based on labels:
+
+```yaml
+- name: microshift-manifests
+  configType: GitConfigProviderSpec
+  gitRef:
+    repository: basic-nginx-demo
+    targetRevision: 1.0.0
+    path: /basic-nginx-demo/configuration/{{ .device.labels["region"] }}/{{ .device.labels["site"] }}
+```
+
+* Enable templating of inline configuration based on device details:
+```yaml
+ - name: motd-update
+   configType: InlineConfigProviderSpec
+   inline:
+     ignition:
+       version: 3.4.0
+     storage:
+       files:
+         - contents:
+             source: >-
+               data:,Device%20{{ .device.name }}%0AThis%20system%20is%20managed%20by%20flightctl.%0A
+           mode: 422
+           overwrite: true
+           path: "/etc/motd"
+```
+### Non-Goals
+
+## Proposal
+
+### Device API
+
+The device API does not change, in the device API we don't allow templating. Any
+device specifications from a fleet should already be rendered into the device
+
+Variable/template expansion is not supported at the device API level, as:
+  * We are operating on a single device, variables aren't really beneficial
+  * Updates to labels will not affect configuration unless the device is part of a fleet.
+
+### Fleet API
+
+The template.spec field of a fleet allows the use of go-template logic to render the device spec.
```

Comment:
Should we document alternatives considered and their pros/cons? And compare to the (IIRC rather limited) use of templating in MCO?

For example, having go-templates here means one cannot process these files with yaml editing tools like `yq` anymore.

An alternative could be a replacement mechanism like in the K8s downward API, but isn't very user-friendly (as in: simple to grok) and I also do not see how we can map the "env vars" back to the fields we want to parameterise.

Reply:
As we spoke about offline. The alternative approach proposed would serve the known device resourse and or provide basic agent templating tokens. Perhaps this is {{ .device.name}} or `__DEVICE_NAME__`  for example. This would not replace a server-side solution but could provide flexibility for files and services looking for this data.

Reply:
@fzdarsky I have modified the examples in a way that it's valid for gotemplate and at the same time it would allow yq. 

Also I am using the index .map `key` which is the right way for gotemplate, the other one is jinja which won't work here :-/

Reply:
I probably didn’t notice: Is your solution to put quotes around the curly braces (`”{{…}}”`)? If so, an alternative could be to change the delimiters to `${{…}}`, which seems to be valid YAML even without quotes.